File: 04_OO/ObjProduct.c
#include <stdio.h>

typedef struct {
    char name[100];
    int stock;
} Product;

int isCorrectQuantity(int quantity) {
    return quantity >= 0;
}
void decreaseStock(Product *p, int quantity) {
    if (isCorrectQuantity(quantity) == 1) { p->stock -= quantity; }
}

void setStock(Product *p, int newStock) {
    if (isCorrectQuantity(newStock) == 1) { p->stock = newStock; }
}

// ログ記録関数
void logInventoryChange(const Product *p, int change) {
    printf("[LOG] %s: %d\n", p->name, change);
}

int main (void) {
    Product laptop = {"Laptop", 50};

    decreaseStock(&laptop, 3);          // 期待通りの動き
    logInventoryChange(&laptop, -3);    // ログも記録

    decreaseStock(&laptop, 2);  // ログ忘れ

    laptop.stock = 48;  // 直接操作：ログ・値の検査なし

    logInventoryChange(&laptop, -999);  // ログだけ変な値が入る
}

File: 04_OO/ObjProduct.java
class Product {
    private String name;
    private int stock;

    public Product(String name, int initialStock) {
        this.name = name;
        this.stock = initialStock;
    }

    private boolean isCorrectQuantity(int quantity) {
        return quantity > 0;
    }
    // 在庫減少：ログ記録を自動実行
    public void decreaseStock(int quantity) throws IllegalArgumentException {
        if (!isCorrectQuantity(quantity)) {
            throw new IllegalArgumentException("数量は1以上です");
        }

        stock -= quantity;
        logInventoryChange(-quantity);
    }

    private void logInventoryChange(int change) {
        System.out.println("[LOG] " + this.name + ": " + change);
    }
}

public class ObjProduct {
    public static void main(String[] args) {
        Product laptop = new Product("Laptop", 50);

        laptop.decreaseStock(3);    // 自動で検査、ログ記録
        laptop.decreaseStock(2);    // 他に書き方がない
    }
}

